### PR TITLE
fix issue #133

### PR DESCRIPTION
**Pull request by craffael**
All MeshBuilders + MeshHierarchy accept now unique_ptrs in the constructor instead of shared_ptrs. In principle this is ready to be merge into master, but I'm not sure whether it will cause any incompatibilities with the exercises/lecturecodes.

### CODE REVIEW

**Comment by hiptmair** [comment]
Please delay the merge until June, when we should prepare a major revision of LehrFEM++. Right now two students do their BSc projects with LehrFEM++ and they clone directly from 'master' .

**Comment by hiptmair** [comment]
Exam is over now. BSc thesis of P. Peters and T. Rohner still ongoing, but both should be able to accommodate small changes.

**Comment by craffael** [comment]
@hiptmair is it ok if I merge this fix? By doing so, all mesh builders will use std::unique_ptr<MeshFactory> from now on which is a bit more consistent :)

File: examples/refinement/refinement_demo.cc [hunk 1]
```diff
@@ -63,9 +63,10 @@ int main(int argc, char **argv) {
   std::cout << "########################################" << std::endl;
   std::cout << "Initialization of data structure for mesh hierarchy"
             << std::endl;
-  std::shared_ptr<lf::mesh::hybrid2d::MeshFactory> mesh_factory_ptr =
-      std::make_shared<lf::mesh::hybrid2d::MeshFactory>(2);
-  lf::refinement::MeshHierarchy multi_mesh(mesh_ptr, mesh_factory_ptr);
+  std::unique_ptr<lf::mesh::hybrid2d::MeshFactory> mesh_factory_ptr =
+      std::make_unique<lf::mesh::hybrid2d::MeshFactory>(2);
+  lf::refinement::MeshHierarchy multi_mesh(mesh_ptr,
```
Comment (hiptmair) [inline examples/refinement/refinement_demo.cc]: User forced to rely on std::move(), but the compile won't allow anything else. Thus, it is safe.

File: lib/lf/refinement/mesh_hierarchy.h [hunk 2]
```diff
@@ -101,8 +101,8 @@ class MeshHierarchy {
    * created by refinement.
    * - Unmarks all edges, see RefineMarked().
    */
-  MeshHierarchy(std::shared_ptr<mesh::Mesh> base_mesh,
-                std::shared_ptr<mesh::MeshFactory> mesh_factory);
+  MeshHierarchy(const std::shared_ptr<mesh::Mesh> &base_mesh,
```
Comment (hiptmair) [inline lib/lf/refinement/mesh_hierarchy.h]: A mesh hierarchy should own a mutable shared pointer to a mesh. Just imagine, a mesh hierarchy is built based on some coarse mesh. Then the shared pointer to the coarse mesh reaches the end of its scope. The coarse mesh must not be deleted, lest the mesh hierarchy lose one of its levels. In my opinion this rules out using a constant shared pointer. I am surprised that the compiler does not complain.

Comment (craffael) [reply lib/lf/refinement/mesh_hierarchy.h]: The mesh hierarchy still owns a shared_ptr to the base_mesh. If you look at Line 36 of https://github.com/craffael/lehrfempp/blob/issue_133/lib/lf/refinement/mesh_hierarchy.cc you can see that a copy of `base_mesh` is pushed onto `meshes_` which is a private member field of the class MeshHierarchy. So it's not possible that the coarsest mesh of a MeshHierarchy gets deleted before the MeshHierarchy itself.

The change from `std::shared_ptr<mesh::Mesh>` to `const std::shared_ptr<mesh::Mesh>&` was actually encouraged by clang-tidy. By doing so we save one unnecessary copy of shared_ptr. I would have to check why clang-tidy didn't suggest this already before.

**Pull request merged**